Sources/SotoCore/Concurrency/ExpiringValue.swift

```diff
@@ -0,0 +1,96 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Soto for AWS open source project
+//
+// Copyright (c) 2023 the Soto project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Soto project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import Logging
+
+/// Type holding a value and an expiration value.
+///
+/// When accessing the value you have to provide a closure that will update the
+/// value if it has expired or is about to expire. The type ensures there is only
+/// ever one value update running at any one time. If an update is already running
+/// when you call `getValue` it will wait on the current update function to finish.
+actor ExpiringValue<T> {
```

Comment:
Does this need to be public?

Reply:
Not at this point. There is a possibility I will use it in some Cognito credential providers. But for the moment I am only using it internally